sdks/python/container/boot.go

```diff
@@ -105,18 +107,57 @@ func main() {
 
 	// (2) Retrieve and install the staged packages.
 
-	dir := filepath.Join(*semiPersistDir, "staged")
+	func() {
 
-	files, err := artifact.Materialize(ctx, *artifactEndpoint, info.GetRetrievalToken(), dir)
-	if err != nil {
-		log.Fatalf("Failed to retrieve staged files: %v", err)
-	}
+        installCompleteFile := filepath.Join(os.TempDir(), "beam.install.complete")
 
-	// TODO(herohde): the packages to install should be specified explicitly. It
-	// would also be possible to install the SDK in the Dockerfile.
-	if setupErr := installSetupPackages(files, dir); setupErr != nil {
-		log.Fatalf("Failed to install required packages: %v", setupErr)
-	}
+		// skip if install already complete
+		_, err = os.Stat(installCompleteFile)
+		if err == nil {
+			return
+		}
+
+		// lock to guard from concurrent artifact retrieval and installation,
+		// when called by child processes in a worker pool
+		lock, err := lockfile.New(filepath.Join(os.TempDir(), "beam.install.lck"))
+		if err != nil {
+			log.Fatalf("Cannot init artifact retrieval lock: %v", err)
+		}
+
+		for err = lock.TryLock(); err != nil; err = lock.TryLock() {
+			switch err {
+			case lockfile.ErrBusy, lockfile.ErrNotExist:
+				time.Sleep(5 * time.Second)
+				log.Printf("Worker %v waiting for artifact retrieval lock: %v", *id, lock)
+			default:
+				log.Fatalf("Worker %v could not obtain artifact retrieval lock: %v", *id, err)
+			}
+		}
+		defer lock.Unlock()
+
+		// skip if install already complete
+		_, err = os.Stat(installCompleteFile)
+		if err == nil {
+			return
```

Comment:
Why is this not within the locked section?

Reply:
Do you want to avoid blocking the workers?

Reply:
This is within the lock section. defer ... above executes after the completion of this function.

I think we should block all workers as intended here, because workers will need the environment to be setup first.

Reply:
Makes sense. I'm new to Go. However, if you look at the above similar check (line 115), there is a possible race condition which can only be caught here where the lock is in place.